File: src-java/src/main/java/SinglyLinkedListArray.java
/**
 * ClassName: SinglyLinkedListArray
 * Package: PACKAGE_NAME
 * Description:
 *
 * @Author AnXin
 * @Create 2026/5/29 20:37
 * @Version 1.0
 */
public class SinglyLinkedListArray {

    private static final int MAX_SIZE = 10000;
    private static final int SENTINEL = 0;

    int[] e = new int[MAX_SIZE];
    int[] ne = new int[MAX_SIZE];
    int idx;
    int size;

    SinglyLinkedListArray() {
        ne[SENTINEL] = -1;
        idx = 1;
        size = 0;
    }

    public void insert(int index, int value) {
        if (index < 0 || index > size) {
            throw new IndexOutOfBoundsException("索引越界");
        }

        int prev = SENTINEL;
        for (int i = 0; i < index; i++) {
            prev = ne[prev];
        }

        e[idx] = value;
        ne[idx] = ne[prev];
        ne[prev] = idx;
        idx++;
        size++;
    }

    public int remove(int index) {
        if (index < 0 || index >= size) {
            throw new IndexOutOfBoundsException("索引越界");
        }

        int prev = SENTINEL;
        for (int i = 0; i < index; i++) {
            prev = ne[prev];
        }

        int delNode = ne[prev];
        ne[prev] = ne[delNode];
        size--;

        return e[delNode];
    }

    public void printSinglyLinkedList() {
        int cur = ne[SENTINEL];
        while (cur != -1) {
            System.out.print(e[cur]);
            if (ne[cur] != -1) {
                System.out.print(" -> ");
            }
            cur = ne[cur];
        }
    }
}

File: src-python/singly_linked_list_array.py
class SinglyLinkedListArray:
    def __init__(self):
        self._MAX_SIZE = 10000
        self._SENTINEL = 0

        self._e: list[int] = [0 for _ in range(0, self._MAX_SIZE)]
        self._ne: list[int] = [0 for _ in range(0, self._MAX_SIZE)]
        self._ne[self._SENTINEL] = -1
        self._idx: int = 1
        self._size: int = 0

    def insert(self, index: int, value: int) -> None:
        if index < 0 or index > self._size:
            raise IndexError("索引越界")

        prev: int = self._SENTINEL
        for _ in range(0, index):
            prev = self._ne[prev]

        self._e[self._idx] = value
        self._ne[self._idx] = self._ne[prev]
        self._ne[prev] = self._idx
        self._idx += 1
        self._size += 1

    def remove(self, index: int) -> int:
        if index < 0 or index >= self._size:
            raise IndexError("索引越界")

        prev: int = self._SENTINEL
        for _ in range(0, index):
            prev = self._ne[prev]

        del_node: int = self._ne[prev]
        self._ne[prev] = self._ne[del_node]
        self._size -= 1

        return self._e[del_node]

    def print_singly_linked_list_array(self) -> None:
        cur: int = self._ne[self._SENTINEL]
        while cur != -1:
            print(self._e[cur], end="")
            if self._ne[cur] != -1:
                print(" -> ", end="")
            cur = self._ne[cur]
        print()

if __name__ == '__main__':
    list = SinglyLinkedListArray()
    list.insert(0, 0)
    list.insert(1, 1)
    list.insert(2, 2)
    list.insert(3, 3)
    list.insert(4, 4)
    list.insert(5, 5)
    list.insert(6, 6)
    list.insert(7, 7)
    list.insert(8, 8)
    list.insert(9, 9)
    list.print_singly_linked_list_array()